File: common/src/main/java/net/impleri/playerskills/server/registry/storage/SkillResourceFile.java
package net.impleri.playerskills.server.registry.storage;

import net.impleri.playerskills.PlayerSkills;
import net.minecraft.server.MinecraftServer;
import net.minecraft.world.level.storage.LevelResource;

import java.io.File;
import java.nio.file.Path;
import java.util.UUID;

class SkillResourceFile {
    private static SkillResourceFile instance;

    public static void createInstance(MinecraftServer server) {
        LevelResource levelResource = new LevelResource(PlayerSkills.MOD_ID);
        Path gameFolder = server.getWorldPath(levelResource);

        instance = new SkillResourceFile(gameFolder);
    }

    public static void destroyInstance() {
        instance = null;
    }

    public static File forPlayer(UUID playerUuid) {
        if (instance == null) {
            throw new RuntimeException("Accessing file when the server is not running");
        }

        return instance.getPlayerFile(playerUuid);
    }

    private static void ensureDirectory(File file) {
        file.mkdirs();
    }

    private final Path storage;

    private SkillResourceFile(Path storagePath) {
        storage = storagePath;
    }

    private File getPlayerFile(UUID playerUuid) {
        return new File(getPlayerDirectory(), playerUuid.toString() + ".skills");
    }

    private File getPlayerDirectory() {
        File playerDir = new File(getStorageDirectory(), "players");
        ensureDirectory(playerDir);

        return playerDir;
    }

    private File getStorageDirectory() {
        File dataDir = storage.toFile();
        ensureDirectory(dataDir);

        return dataDir;
    }
}

File: common/src/main/scala/net/impleri/playerskills/server/skills/storage/SkillResourceFile.scala
package net.impleri.playerskills.server.skills.storage

import net.impleri.playerskills.facades.minecraft.Server
import org.jetbrains.annotations.VisibleForTesting

import java.io.File
import java.nio.file.Path
import java.util.UUID
import scala.util.chaining.scalaUtilChainingOps

/**
 * Manages _where_ to save data
 */
class SkillResourceFile private[skills] (private[skills] val storage: Path) {
  private def storageDirectory: File = {
    storage.toFile
      .tap(_.mkdirs())
  }

  private def playerDirectory: File = {
    storageDirectory
      .pipe(new File(_, "players"))
      .tap(_.mkdirs())
  }

  // package-private as this should be accessed through SkillStorage
  private[skills] def getPlayerFile(playerId: UUID): File = {
    playerDirectory
      .pipe(new File(_, s"$playerId.skills"))
  }
}

object SkillResourceFile {
  @VisibleForTesting
  private[skills] def apply(storage: Path): SkillResourceFile = {
    new SkillResourceFile(storage)
  }

  protected[skills] def apply(server: Server): SkillResourceFile = {
    apply(server.getWorldPath())
  }
}
